File: finalyearproject/repos/dkd-platform-libs/sdk/python/src/dkd_platform/dlq.py
# HAND-AUTHORED platform primitive (NOT dkdgen-generated).
# PL-02 — dead-letter queue with per-aggregate-key park-and-freeze (SA-MSG-09/10).
"""Dead-letter queue: a poison money/custody/inventory event parks ONLY its own aggregate key.

Canonical schema (the SDK standard):

    dlq(id BIGSERIAL PK, event_id TEXT, topic TEXT, key TEXT, payload JSONB, error TEXT,
        parked_at TIMESTAMPTZ default now(), aggregate_key TEXT)

Park-and-freeze (SA-MSG-10): a poison message is quarantined under its `aggregate_key`, never
silently dropped. `is_key_parked` lets a consumer freeze further progress on that single key while
every other key keeps flowing.
"""
from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any

from .dbexec import DbExecutor

_INSERT = (
    'INSERT INTO dlq(event_id, topic, "key", payload, error, aggregate_key, parked_at) '
    "VALUES (%s, %s, %s, %s::jsonb, %s, %s, now())"
)
_SELECT_KEY = "SELECT 1 FROM dlq WHERE aggregate_key = %s LIMIT 1"


def _json_param(payload: Any) -> str:
    if isinstance(payload, str):
        return payload
    return json.dumps(payload, sort_keys=True, separators=(",", ":"))


@dataclass(frozen=True)
class DlqEntry:
    """The park input — `{eventId, topic, key, payload, error, aggregateKey}` in Python casing."""

    event_id: str
    topic: str
    key: str
    payload: Any
    error: str
    aggregate_key: str


class Dlq:
    @staticmethod
    def park(db: DbExecutor, entry: DlqEntry) -> None:
        """Quarantine a poison event under its aggregate key (never drop it)."""
        db.execute(
            _INSERT,
            (
                entry.event_id,
                entry.topic,
                entry.key,
                _json_param(entry.payload),
                entry.error,
                entry.aggregate_key,
            ),
        )

    @staticmethod
    def is_key_parked(db: DbExecutor, aggregate_key: str) -> bool:
        """True when a poison event is frozen on this aggregate key (progress must halt for it)."""
        cur = db.execute(_SELECT_KEY, (aggregate_key,))
        return cur.fetchone() is not None

File: finalyearproject/repos/dkd-finance/sdk/src/main/java/com/dokandar/platform/Dlq.java
// HAND-AUTHORED platform primitive (NOT dkdgen-generated).
// PL-02 — dead-letter quarantine with per-aggregate-key "park-and-freeze" (SA-MSG-09/10).
// A poison money/custody/inventory event parks ONLY its own aggregate key: that key freezes
// while every other key on the topic keeps progressing. Poison events are never silently dropped.
package com.dokandar.platform;

import java.util.List;
import java.util.Map;

/**
 * Dead-letter queue + park-and-freeze gate. When an event repeatedly fails, {@link #park} records it
 * (never drops it) and freezes its {@code aggregateKey}; the consumer calls {@link #isKeyParked}
 * before processing so it skips any further event on a frozen key while other keys flow (SA-MSG-10).
 * Replay/unpark is an operational four-eyes action outside this helper.
 */
public final class Dlq {

    private Dlq() {
    }

    /** Canonical DLQ DDL — the SDK standard for the quarantine sink. */
    public static final String DDL = """
        CREATE TABLE IF NOT EXISTS dlq (
          id            BIGSERIAL PRIMARY KEY,
          event_id      TEXT        NOT NULL,
          topic         TEXT        NOT NULL,
          key           TEXT        NOT NULL,
          payload       JSONB       NOT NULL,
          error         TEXT        NOT NULL,
          aggregate_key TEXT        NOT NULL,
          parked_at     TIMESTAMPTZ NOT NULL DEFAULT now()
        );
        CREATE INDEX IF NOT EXISTS dlq_aggregate_key_idx ON dlq (aggregate_key);""";

    private static final String INSERT_SQL =
        "INSERT INTO dlq (event_id, topic, key, payload, error, aggregate_key, parked_at) "
        + "VALUES (?, ?, ?, ?::jsonb, ?, ?, now())";

    private static final String IS_PARKED_SQL =
        "SELECT 1 FROM dlq WHERE aggregate_key = ? LIMIT 1";

    /** A poison event being quarantined (freezes {@code aggregateKey}). */
    public record Record(String eventId, String topic, String key, String payload,
                         String error, String aggregateKey) {
        public Record {
            req(eventId, "eventId");
            req(topic, "topic");
            req(key, "key");
            req(payload, "payload");
            req(aggregateKey, "aggregateKey");
        }
    }

    /** Quarantine {@code rec} and freeze its aggregate key. Idempotent replay is safe (append-only). */
    public static void park(SqlExecutor db, Record rec) {
        db.execute(INSERT_SQL, rec.eventId(), rec.topic(), rec.key(),
            rec.payload(), rec.error(), rec.aggregateKey());
    }

    /** True when {@code aggregateKey} has a parked poison event (skip further events on that key). */
    public static boolean isKeyParked(SqlExecutor db, String aggregateKey) {
        req(aggregateKey, "aggregateKey");
        List<Map<String, Object>> rows = db.execute(IS_PARKED_SQL, aggregateKey);
        return !rows.isEmpty();
    }

    private static void req(String v, String field) {
        if (v == null || v.isBlank()) {
            throw new IllegalArgumentException("dlq: " + field + " must be non-blank");
        }
    }
}
